src/dmd/cparse.d

```diff
@@ -0,0 +1,3373 @@
+            check(TOK.leftParenthesis);
+            auto condition = cparseExpression();
+            check(TOK.rightParenthesis);
+            check(TOK.semicolon, "terminating `;` required after do-while statement");
+            s = new AST.DoStatement(loc, _body, condition, token.loc);
+            break;
+        }
+
+        case TOK.for_:
+        {
+            AST.Statement _init;
+            AST.Expression condition;
+            AST.Expression increment;
+
+            nextToken();
+            check(TOK.leftParenthesis);
+            if (token.value == TOK.semicolon)
+            {
+                _init = null;
+                nextToken();
+            }
+            else
+            {
+                _init = cparseStatement(0);
+            }
+            if (token.value == TOK.semicolon)
+            {
+                condition = null;
+                nextToken();
+            }
+            else
+            {
+                condition = cparseExpression();
+                check(TOK.semicolon, "`for` condition");
+            }
+            if (token.value == TOK.rightParenthesis)
+            {
+                increment = null;
+                nextToken();
+            }
+            else
+            {
+                increment = cparseExpression();
+                check(TOK.rightParenthesis);
+            }
+            Loc endloc;
+            auto _body = cparseStatement(ParseStatementFlags.scope_, null, &endloc);
+            s = new AST.ForStatement(loc, _init, condition, increment, _body, endloc);
+            break;
+        }
+
+        case TOK.if_:
+        {
+            nextToken();
+            check(TOK.leftParenthesis);
+            auto condition = cparseExpression();
+            check(TOK.rightParenthesis);
+            auto ifbody = cparseStatement(ParseStatementFlags.scope_);
+            AST.Statement elsebody;
+            if (token.value == TOK.else_)
+            {
+                nextToken();
+                elsebody = cparseStatement(ParseStatementFlags.scope_);
+            }
+            else
+                elsebody = null;
+            if (condition && ifbody)
+                s = new AST.IfStatement(loc, null, condition, ifbody, elsebody, token.loc);
+            else
+                s = null; // don't propagate parsing errors
+            break;
+        }
+
+        case TOK.else_:
+            error("found `else` without a corresponding `if` statement");
+            goto Lerror;
+
+        case TOK.switch_:
+        {
+            nextToken();
+            check(TOK.leftParenthesis);
+            auto condition = cparseExpression();
+            check(TOK.rightParenthesis);
+            auto _body = cparseStatement(ParseStatementFlags.scope_);
+            s = new AST.SwitchStatement(loc, condition, _body, false);
+            break;
+        }
+
+        case TOK.case_:
+        {
+
+            nextToken();
+            auto exp = parseAssignExp();
```

Comment:
I just noticed: this should be `cparseAssignExp`. Now it's actually parsing a D assign expression inherited from `Parser`.

Reply:
https://github.com/dlang/dmd/pull/12531